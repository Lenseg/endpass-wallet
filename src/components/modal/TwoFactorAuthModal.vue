<template lang="html">
  <v-modal @close="close">
    <template slot="header">Two Factor Authentication via OTP</template>

    <v-form @submit="confirm">
      <div class="field" v-if="secret && email">
        <div class="has-text-centered">
          <img :src="qrCodeSrc" />
        </div>
        <label class="label">Secret</label>
        {{secret}}
      </div>
      <v-input v-model="code"
               label="Verification Code"
               name="verificationCode"
               validator="required"/>
      <v-button className="is-primary is-medium" :loading="isLoading">Verify</v-button>
    </v-form>
  </v-modal>
</template>

<script>
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VButton from '@/components/ui/form/VButton';

export default {
  name: 'two-factor-auth-modal',
  data() {
    return {
      code: null,
    };
  },
  computed: {
    qrCodeSrc() {
      const otpAuthUri = `otpauth://totp/Endpass:${this.email}?secret=${
        this.secret
      }`;

      return `https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=${otpAuthUri}`;
    },
  },
  props: {
    secret: String,
    email: String,
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    confirm() {
      this.$emit('confirm', this.code);
    },
    close() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VForm,
    VInput,
    VButton,
  },
};
</script>
