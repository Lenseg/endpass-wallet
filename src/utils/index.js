export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());
