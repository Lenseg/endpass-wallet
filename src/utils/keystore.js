import keythereum from 'keythereum';
import bs58check from 'bs58check';
import { kdfParams } from '@/config';

// Monkey patch keythereum to skip generating address for private keys
// This allows us to encrypt private keys of arbitrary length, and
// conforms better to the Ethereum keystore V3 spec, which does not include
// the address for privacy reasons.
// See https://github.com/ethereum/wiki/wiki/Web3-Secret-Storage-Definition#alterations-from-version-1
keythereum.privateKeyToAddress = function(pk) {
  return '';
};

export default {
  // Encrypts a private key Buffer into a V3 keystore object
  // The exported keystore does NOT include an address
  encrypt(password, privateKey) {
    // Generate random salt and iv for each encryption
    let dk = keythereum.create();
    let options = {
      kdf: kdfParams.kdf,
      kdfparams: kdfParams,
    };
    let encrypted = keythereum.dump(
      password,
      privateKey,
      dk.salt,
      dk.iv,
      options,
    );
    delete encrypted.address;
    return encrypted;
  },

  // Decrypts a V3 keystore object into a private key Buffer
  decrypt(password, keystore) {
    return keythereum.recover(password, keystore);
  },

  // Encode to Base58Check
  encodeBase58(key) {
    return bs58check.encode(key);
  },

  // Decode from Base58Check
  decodeBase58(key) {
    return bs58check.decode(key);
  },
};
