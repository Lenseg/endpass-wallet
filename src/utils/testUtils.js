const toKebab = string =>
  string.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let dashName;

    if (stubComponent.name) {
      dashName = toKebab(stubComponent.name);
    }

    stubs[dashName] = {
      render(createElement) {
        return createElement(dashName, this.$slots.default);
      },
    };

    return stubs;
  }, {});
}
