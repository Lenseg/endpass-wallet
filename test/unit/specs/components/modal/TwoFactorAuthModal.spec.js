import { shallow } from '@vue/test-utils';
import TwoFactorAuthModal from '@/components/modal/TwoFactorAuthModal';
import { generateStubs } from '@/utils/testUtils';

describe('TwoFactorAuthModal', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(TwoFactorAuthModal, {
      stubs: generateStubs(TwoFactorAuthModal),
    });
  });

  describe('render', () => {
    it('should render without secret and QR code', () => {
      expect(wrapper.element).toMatchSnapshot();

      wrapper.setProps({
        email: null,
        secret: 'secret',
      });

      expect(wrapper.element).toMatchSnapshot();

      wrapper.setProps({
        email: 'email',
        secret: null,
      });

      expect(wrapper.element).toMatchSnapshot();
    });

    it('should render with secret and QR code', () => {
      wrapper.setProps({
        email: 'email',
        secret: 'secret',
      });

      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
