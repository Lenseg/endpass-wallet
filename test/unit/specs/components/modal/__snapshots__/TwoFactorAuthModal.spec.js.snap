// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TwoFactorAuthModal render should render with secret and QR code 1`] = `
<v-modal>
   
  <v-form>
    <div
      class="field"
    >
      <div
        class="has-text-centered"
      >
        <img
          src="https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=otpauth://totp/Endpass:email?secret=secret"
        />
      </div>
       
      <label
        class="label"
      >
        Secret
      </label>
      
      secret
    
    </div>
     
    <v-input
      label="Verification Code"
      validator="required"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;

exports[`TwoFactorAuthModal render should render without secret and QR code 1`] = `
<v-modal>
   
  <v-form>
    <!---->
     
    <v-input
      label="Verification Code"
      validator="required"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;

exports[`TwoFactorAuthModal render should render without secret and QR code 2`] = `
<v-modal>
   
  <v-form>
    <!---->
     
    <v-input
      label="Verification Code"
      validator="required"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;

exports[`TwoFactorAuthModal render should render without secret and QR code 3`] = `
<v-modal>
   
  <v-form>
    <!---->
     
    <v-input
      label="Verification Code"
      validator="required"
    />
     
    <v-button
      classname="is-primary is-medium"
    >
      Verify
    </v-button>
  </v-form>
</v-modal>
`;
